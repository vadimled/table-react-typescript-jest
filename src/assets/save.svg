<svg width="14" height="14" viewBox="0 0 14 14" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M4.24821 2.5044H9.31136C9.46201 2.5044 9.58542 2.38143 9.58542 2.23046V0H3.97363V2.23046C3.97363 2.38143 4.09671 2.5044 4.24821 2.5044Z" fill="#999FA3"/>
<path d="M13.3464 3.41044L10.4891 0.541112V2.23043C10.4891 2.88007 9.9605 3.40867 9.31085 3.40867H4.24771C3.59792 3.40867 3.06947 2.88007 3.06947 2.23043V0H1.24414C0.55699 0 0 0.557101 0 1.24378V12.3145C0 13.0011 0.55699 13.5583 1.24414 13.5583H12.3143C13.0011 13.5583 13.5581 13.0012 13.5581 12.3145V3.92288C13.5581 3.73044 13.4819 3.54629 13.3464 3.41044ZM11.449 10.7053C11.449 11.1059 11.1241 11.4313 10.7226 11.4313H2.83505C2.43392 11.4313 2.1091 11.1059 2.1091 10.7053V6.92012C2.1091 6.51958 2.43392 6.19431 2.83505 6.19431H10.7226C11.1241 6.19431 11.449 6.51963 11.449 6.92012V10.7053Z" fill="#999FA3"/>
<path d="M9.89523 7.3595H3.66225C3.41221 7.3595 3.20996 7.56138 3.20996 7.8114C3.20996 8.06141 3.41221 8.26329 3.66225 8.26329H9.89525C10.1447 8.26329 10.3471 8.06141 10.3471 7.8114C10.3471 7.56138 10.1448 7.3595 9.89523 7.3595Z" fill="#999FA3"/>
<path d="M8.99154 9.16748H4.56616C4.3165 9.16748 4.11426 9.36936 4.11426 9.61938C4.11426 9.86889 4.3165 10.0713 4.56616 10.0713H8.99151C9.24117 10.0713 9.44344 9.86889 9.44344 9.61938C9.44347 9.36936 9.2412 9.16748 8.99154 9.16748Z" fill="#999FA3"/>
</svg>
