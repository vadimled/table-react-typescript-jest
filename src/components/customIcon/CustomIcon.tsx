/**
 The CustomIcon component renders the any Icon transmitted through props as a ReactComponent.
 If the property "disabled" is false then either the Icon or the Spinner is drawn
 If the property "disabled" is true then then the Icon will drawn with gray color
 */
import React, { FC, SVGProps } from "react";
import { Spin } from "antd";
import { SyncOutlined } from "@ant-design/icons";
import * as styles from "./CustomIcon.module.scss";
import cn from "classnames";

interface CustomIconType {
  disabled?: boolean;
  Icon: FC<SVGProps<SVGSVGElement>>;
  loading?: boolean;
  onClick: (e: React.MouseEvent) => void;
}

const CustomIcon: FC<CustomIconType> = ({
  disabled,
  loading,
  Icon,
  onClick,
}) => {
  return (
    <div
      className={styles["custom-icon-wrapper"]}
      onClick={onClick}
      data-testid="custom-icon"
    >
      {loading && !disabled ? (
        <Spin
          data-testid="antd-spinner"
          indicator={
            <SyncOutlined data-testid="icon-spin" className="spinner" spin />
          }
        />
      ) : (
        <div
          data-testid="target-icon"
          className={cn("current-icon", { disabled: disabled })}
        >
          <Icon />
        </div>
      )}
    </div>
  );
};

export default CustomIcon;
