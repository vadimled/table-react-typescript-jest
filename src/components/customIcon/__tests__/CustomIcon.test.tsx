import React from "react";
import CustomIcon from "../CustomIcon";
import { ReactComponent as SaveIcon } from "../../../assets/save.svg";
import { fireEvent, render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

describe("CustomIcon", (): void => {
  afterEach(cleanup);
  it("snapshot of component view if 'disabled===false && loading===false' ", (): void => {
    const { asFragment } = render(
      <CustomIcon Icon={SaveIcon} onClick={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("should have been called ones after click", (): void => {
    const onIconClick = jest.fn();
    const { getByTestId } = render(
      <CustomIcon Icon={SaveIcon} onClick={onIconClick} />
    );
    const customIconWrapper = getByTestId("custom-icon");
    fireEvent.click(customIconWrapper);
    expect(onIconClick).toHaveBeenCalledTimes(1);
  });
  it("element should has class 'disabled' if 'disabled===true' ", (): void => {
    const { getByTestId } = render(
      <CustomIcon Icon={SaveIcon} onClick={jest.fn()} disabled />
    );
    const targetIcon = getByTestId("target-icon");
    expect(targetIcon).toHaveClass("disabled");
  });
  it("The 'antd-spinner' should not exists if component property 'disabled' is true", (): void => {
    const { queryByTestId } = render(
      <CustomIcon Icon={SaveIcon} onClick={jest.fn()} disabled />
    );
    expect(queryByTestId("antd-spinner")).toBeNull();
  });
});
