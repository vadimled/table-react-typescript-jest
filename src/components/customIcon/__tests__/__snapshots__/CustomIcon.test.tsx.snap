// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomIcon snapshot of component view if 'disabled===false && loading===false'  1`] = `
<DocumentFragment>
  <div
    data-testid="custom-icon"
  >
    <div
      class="current-icon"
      data-testid="target-icon"
    >
      <reactcomponent />
    </div>
  </div>
</DocumentFragment>
`;
